cmd/get_client_certificate.go

```diff
@@ -156,71 +161,79 @@ func NewCmdGetClientCertificate(f util.Factory, ioStreams genericclioptions.IOSt
-	obj, _, err := exec.LoadExecCredentialFromEnv()
-	if err != nil {
-		return err
-	}
+	env := os.Getenv(execInfoEnv)
+	if env != "" { // KUBERNETES_EXEC_INFO env variable set for kubectl versions starting with v1.20.0
+		obj, _, err := exec.LoadExecCredential([]byte(env))
 
-	cred, ok := obj.(*clientauthv1beta1.ExecCredential)
-	if !ok {
-		return fmt.Errorf("cannot convert to ExecCredential: %w", err)
-	}
+		if err != nil {
+			return err
+		}
 
-	var extension ExecPluginConfig
+		cred, ok := obj.(*clientauthv1beta1.ExecCredential)
+		if !ok {
+			return fmt.Errorf("cannot convert to ExecCredential: %w", err)
+		}
 
-	if cred.Spec.Cluster.Config.Raw != nil {
-		if err := json.Unmarshal(cred.Spec.Cluster.Config.Raw, &extension); err != nil {
-			return err
+		var extension ExecPluginConfig
+
+		if cred.Spec.Cluster.Config.Raw != nil {
+			if err := json.Unmarshal(cred.Spec.Cluster.Config.Raw, &extension); err != nil {
+				return err
+			}
 		}
-	}
 
-	o.ShootCluster = cred.Spec.Cluster
+		o.ShootCluster = cred.Spec.Cluster
 
-	if o.GardenClusterIdentity == "" {
-		o.GardenClusterIdentity = extension.GardenClusterIdentity
-	}
+		if o.GardenClusterIdentity == "" {
+			o.GardenClusterIdentity = extension.GardenClusterIdentity
+		}
 
-	if o.ShootRef.Name == "" {
-		o.ShootRef.Name = extension.ShootRef.Name
-	}
+		if o.ShootRef.Name == "" {
+			o.ShootRef.Name = extension.ShootRef.Name
+		}
 
-	if o.ShootRef.Namespace == "" {
-		o.ShootRef.Namespace = extension.ShootRef.Namespace
+		if o.ShootRef.Namespace == "" {
+			o.ShootRef.Namespace = extension.ShootRef.Namespace
+		}
 	}
 
-	o.CertificateCacheStore = &store.Store{Dir: o.CertificateCacheDir}
+	o.CertificateCacheStore = f.CertificateStore(o.CertificateCacheDir)
 
-	o.GardenCoreV1Beta1RESTClient, err = f.RESTClient(o.GardenClusterIdentity)
-	if err != nil {
-		return err
+	var err error
+
+	if o.GardenClusterIdentity != "" {
+		o.GardenCoreV1Beta1RESTClient, err = f.RESTClient(o.GardenClusterIdentity)
+		if err != nil {
+			return err
+		}
 	}
 
+	o.Clock = f.Clock()
+
 	return nil
 }
 
 // Validate makes sure provided values for GetClientCertificateOptions are valid
 func (o *GetClientCertificateOptions) Validate() error {
-	if o.ShootCluster == nil {
-		return errors.New("cluster must be specified")
-	}
-
-	if len(o.ShootCluster.Server) == 0 {
-		return errors.New("server must be specified")
-	}
+	if os.Getenv(execInfoEnv) != "" {
+		if o.ShootCluster == nil {
+			return errors.New("cluster must be specified")
+		}
 
-	if len(o.ShootCluster.CertificateAuthorityData) == 0 {
-		return errors.New("certificate authority data must be specified")
+		if len(o.ShootCluster.Server) == 0 {
+			return errors.New("server must be specified")
+		}
 	}
 
 	if len(o.ShootRef.Name) == 0 {
-		return errors.New("name must be specified")
+		return errors.New("name must be specified. Hint: update kubectl in case you are using a version older than v1.20.0")
```

Comment:
Why is this hint only relevant if no name is given? Does it not apply also to the 2 if statements down from here?

Reply:
Yes in theory it would apply also for the other 2 if statements below, but as it would return with an error on this if statement I did not see a point to add the hint to the other if statements below.